rest/route/project.go

```diff
@@ -185,3 +186,133 @@ func (h *versionsGetHandler) Run(ctx context.Context) gimlet.Responder {
 
 	return gimlet.NewJSONResponse(versions)
 }
+
+/*
+ * Creates project (project_ref)
+ * PUT: /rest/v2/projects/
+ * Perm: superUser
+ */
+type projectCreateHandler struct {
+	projectRef *model.APIProjectRef
+
+	sc data.Connector
+}
+
+func makeProjectCreateRoute(sc data.Connector) gimlet.RouteHandler {
+	return &projectCreateHandler{sc: sc}
+}
+
+func (h *projectCreateHandler) Factory() gimlet.RouteHandler {
+	return &projectCreateHandler{sc: h.sc}
+}
+
+func (h *projectCreateHandler) Parse(ctx context.Context, r *http.Request) error {
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	if err := util.ReadJSONInto(body, &h.projectRef); err != nil {
+		return errors.Wrap(err, "Cannot parse the content as JSON!")
+	}
+
+	return nil
+}
+
+func (h *projectCreateHandler) Run(ctx context.Context) gimlet.Responder {
+	createdProject, err := h.sc.CreateProject(h.projectRef)
+
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "Cannot create project!"))
+	}
+
+	apiProject := &model.APIProject{}
+	// The method doesn't accept pointers, we have to dereference it first
+	err = apiProject.BuildFromService(*createdProject)
+
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			Message:    "problem converting project document",
+			StatusCode: http.StatusInternalServerError,
+		})
```

Comment:
This error doesn't propagate information from the caller, (i.e. the current value of err). There are a couple of solutions: 

- MakeJSONErrorResponder (and friends, including MakeJSONInternalErrorResponder) will convert a go error into a gimlet.ErrorResponse object (using the default value from the function type, and putting the value of err.Error() as the message. This is often what you want. using errors.wrap here is fine. 
- You can have BuildFromService return a gimlet.ErrorResponse object, which implements go's error interface, and then you don't need to wrap it here (though you could.) 
- Some combination of the above.

Reply:
Regarding №2, in my opinion, api models shouldn't rely on API details implementation. So I would go for №1.